source/httpfs/driver_test.go

```diff
@@ -0,0 +1,101 @@
+package httpfs
+
+import (
+	"net/http"
+	"testing"
+
+	st "github.com/golang-migrate/migrate/v4/source/testing"
+)
+
+func TestWithInstanceAndNew(t *testing.T) {
+	tests := []struct {
+		msg  string
+		fs   http.FileSystem
+		path string
+		ok   bool
+	}{
+		{
+			msg: "valid dir, empty path",
+			fs:  http.Dir("testdata/sql"),
+			ok:  true,
+		},
+		{
+			msg:  "valid dir, non-empty path",
+			fs:   http.Dir("testdata"),
+			path: "sql",
+			ok:   true,
+		},
+		{
+			msg: "invalid dir",
+			fs:  http.Dir("does-not-exist"),
+		},
+		{
+			msg: "file instead of dir",
+			fs:  http.Dir("testdata/sql/1_foobar.up.sql"),
+		},
+		{
+			msg: "dir with duplicates",
+			fs:  http.Dir("testdata/duplicates"),
+		},
+	}
+
+	for _, test := range tests {
+		d, err := WithInstance(test.fs, test.path)
```

Comment:
Reason for not using sub-tests here is to avoid adding test names for all the test cases. If we do something like:

```go
        for _, test := range tests {
                t.Run(test.msg, func(t *testing.T) {
                        ...
                })
        }
```

we would have cryptic test names:

```
=== RUN   TestWithInstanceAndNew/valid_dir,_empty_path
=== RUN   TestWithInstanceAndNew/valid_dir,_non-empty_path
=== RUN   TestWithInstanceAndNew/invalid_dir
=== RUN   TestWithInstanceAndNew/file_instead_of_dir
=== RUN   TestWithInstanceAndNew/dir_with_duplicates
--- PASS: TestWithInstanceAndNew (0.00s)
    --- PASS: TestWithInstanceAndNew/valid_dir,_empty_path (0.00s)
    --- PASS: TestWithInstanceAndNew/valid_dir,_non-empty_path (0.00s)
    --- PASS: TestWithInstanceAndNew/invalid_dir (0.00s)
    --- PASS: TestWithInstanceAndNew/file_instead_of_dir (0.00s)
    --- PASS: TestWithInstanceAndNew/dir_with_duplicates (0.00s)
```

Of course we could simplify test case messages to work nicely as sub-test names. But then we would loose descriptive test case definitions and gain potentially nothing in return.

We could add sub-test names in addition to test case definitions.

Which approach would you prefer?

Reply:
Renaming `msg` to `name` and dropping the usage of the string in the error messages would help. The string values seem fine to me as is.